packages/swarm-components/src/Dropdown.jsx

```diff
@@ -0,0 +1,453 @@
+	Children,
+	useContext,
+	useState,
+	useEffect,
+	useRef,
+} from 'react';
+import Portal from '@reach/portal';
+import Rect, { useRect } from '@reach/rect';
+import WindowSize from '@reach/window-size';
+import { node, func, object, string, number, oneOfType, any } from 'prop-types';
+import * as ReachUtils from '@reach/utils';
+import { ForwardedButton } from './Button';
+
+// TODO: add the mousedown/drag/mouseup to select of native menus, will
+// also help w/ remove the menu button tooltip hide-flash.
+
+// TODO: add type-to-highlight like native menus
+
+const MenuContext = createContext();
+
+const checkIfAppManagedFocus = ({ refs, state }) =>
+	state.isOpen && Boolean(refs.menu) && refs.menu.contains(document.activeElement);
+
+const openAtFirstItem = state => ({ isOpen: true, selectionIndex: 0 });
+
+const closeState = state => ({
+	isOpen: false,
+	selectionIndex: -1,
+	closingWithClick: false,
+});
+
+const genId = prefix =>
+	`${prefix}-${Math.random()
+		.toString(32)
+		.substr(2, 8)}`;
+
+////////////////////////////////////////////////////////////////////////
+
+const INITIAL_MENU_STATE = {
+	// isOpen: false,
+	buttonRect: {},
+	selectionIndex: -1,
+	closingWithClick: false,
+	buttonId: genId('button'),
+};
+
+const checkIfStylesIncluded = () => ReachUtils.checkStyles('menu-button');
+
+const Menu = ({ children }) => {
+	const [state, setState] = useState(INITIAL_MENU_STATE);
+	const [appManagedFocus, setAppManagedFocus] = useState(false);
+
+	const context = {
+		state,
+		setState,
+		refs: {
+			button: null,
+			menu: null,
+			items: [],
+		},
+	};
+
+	useEffect(() => {
+		checkIfStylesIncluded();
+		setAppManagedFocus(checkIfAppManagedFocus(context));
+	}, [state.isOpen]);
+
+	return (
+		<MenuContext.Provider value={context}>
+			{typeof children === 'function'
+				? children({ isOpen: context.state.isOpen })
+				: children}
+		</MenuContext.Provider>
+	);
+};
+
+Menu.propTypes = {
+	children: oneOfType([func, node]),
+};
+
+////////////////////////////////////////////////////////////////////////
+const MenuButton = React.forwardRef(
+	({ onClick, onKeyDown, onMouseDown, ...props }, ref) => {
+		const { refs, state, setState } = useContext(MenuContext);
+		const buttonRef = useRef(null);
+		const buttonRect = useRect(buttonRef, state.isOpen);
+
+		useEffect(() => {
+			if (!state.isOpen) {
+				buttonRef.current.focus();
+			}
+		}, [state.isOpen]);
+
+		useEffect(() => {
+			setState({ ...state, buttonRect });
+		}, [state.isOpen, buttonRect]);
+
+		return (
+			<ForwardedButton
+				id={state.buttonId}
+				aria-haspopup="menu"
+				aria-expanded={state.isOpen}
+				data-reach-menu-button
+				type="button"
+				ref={buttonRef}
+				onMouseDown={ReachUtils.wrapEvent(onMouseDown, () => {
+					if (state.isOpen) {
+						setState({ ...state, closingWithClick: true });
+					}
+				})}
+				onClick={ReachUtils.wrapEvent(onClick, () => {
+					if (state.isOpen) {
+						setState({ ...state, ...closeState() });
```

Comment:
i guess i'm just not used to this, but what is the purpose of having a function in state? also what exactly is being set here?